### PR TITLE
Initializes the restrictions list for each network

### PR DESCRIPTION
**Pull request by vicsn**
## Motivation

Mirror of: https://github.com/ProvableHQ/snarkVM/pull/15
Passed review and CI.

### CODE REVIEW

**Review by Meshiest — APPROVED** [review]
These changes look good assuming they are used with caution. I suspect if a restriction is added retroactively (on block ranges that have already been added to the ledger), it will corrupt the mappings for newly syncing nodes and result in a desync between those and currently synced nodes.

**Comment by vicsn** [comment]
> I suspect if a restriction is added retroactively (on block ranges that have already been added to the ledger), it will corrupt the mappings for newly syncing nodes and result in a desync between those and currently synced nodes.

Caution indeed advised! But the above scenario is not possible because the node with updated list will not connect and sync blocks from nodes with other lists: https://github.com/AleoNet/snarkOS/pull/3306

File: console/network/src/canary_v0.rs [hunk 1]
```diff
@@ -152,6 +152,11 @@ impl Network for CanaryV0 {
         snarkvm_parameters::canary::GenesisBytes::load_bytes()
     }
 
+    /// Returns the restrictions list as a JSON-compatible string.
+    fn restrictions_list_as_str() -> &'static str {
```
Comment (evanmarshall) [inline console/network/src/canary_v0.rs:156]: I'm not sure about the naming here. I think I'd prefer `restrictions_list()` or `restrictions_list_json()` to be more clear about the intent.

File: console/network/src/lib.rs [hunk 2]
```diff
@@ -215,6 +215,9 @@ pub trait Network:
     /// Returns the genesis block bytes.
     fn genesis_bytes() -> &'static [u8];
 
+    /// Returns the restrictions list as a JSON-compatible string.
+    fn restrictions_list_as_str() -> &'static str;
```
Comment (evanmarshall) [inline console/network/src/lib.rs:219]: Same comment here about the naming.

File: console/network/src/mainnet_v0.rs [hunk 3]
```diff
@@ -149,6 +149,11 @@ impl Network for MainnetV0 {
         snarkvm_parameters::mainnet::GenesisBytes::load_bytes()
     }
 
+    /// Returns the restrictions list as a JSON-compatible string.
+    fn restrictions_list_as_str() -> &'static str {
```
Comment (evanmarshall) [inline console/network/src/mainnet_v0.rs:153]: Same here about the naming

File: console/network/src/testnet_v0.rs [hunk 4]
```diff
@@ -152,6 +152,11 @@ impl Network for TestnetV0 {
         snarkvm_parameters::testnet::GenesisBytes::load_bytes()
     }
 
+    /// Returns the restrictions list as a JSON-compatible string.
+    fn restrictions_list_as_str() -> &'static str {
```
Comment (evanmarshall) [inline console/network/src/testnet_v0.rs:156]: Same here about the naming

File: ledger/benches/bonded_mapping.rs [hunk 5]
```diff
@@ -22,12 +22,13 @@ use console::{
     program::{Identifier, Literal, Plaintext, ProgramID, Value},
 };
 use ledger_committee::{MAX_DELEGATORS, MIN_DELEGATOR_STAKE};
+use ledger_store::ConsensusStore;
+use synthesizer::VM;
```
Comment (evanmarshall) [inline ledger/benches/bonded_mapping.rs:26]: Should we add dependency ordering to the pre-commit hooks? Not here but perhaps as a follow up issue.

File: synthesizer/src/restrictions/mod.rs [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+
+impl<N: Network> Restrictions<N> {
+    /// Initializes the `Restrictions` instance for the current network.
+    pub fn load() -> Result<Self> {
+        // Load the restrictions list from the network.
+        let restrictions = Self::from_str(N::restrictions_list_as_str())?;
+        // Ensure the restrictions ID matches the computed value.
+        let expected_restrictions_id =
+            Self::compute_restrictions_id(&restrictions.programs, &restrictions.functions, &restrictions.arguments)?;
+        if restrictions.restrictions_id != expected_restrictions_id {
```
Comment (evanmarshall) [inline synthesizer/src/restrictions/mod.rs:69]: It seems like we could leave the `restrictions_id` out of the json serialized file. It seems like we need either: a script to generate the restrictions_id from `programs`, `functions`, and `arguments` or we could just omit the `restrictions_id` from the json and calculate it.

File: synthesizer/src/restrictions/mod.rs [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+    /// e.g. `restricted.aleo/bar _ aleo1zkpxxxxx _ _` => `..10` (up to block 10)
+    /// e.g. `restricted.aleo/bar _ aleo1zkpxxxxx _ _` => `10..20` (from block 10 to block 20)
+    ///
+    /// Note: This design intentionally minimizes the number of total lookups required to check
+    /// for restrictions when a transition matches the described profile. In summary:
+    /// - When a transition does not match the program ID or function name, the total lookup cost is `O(1)`.
+    /// - When a transition matches the program ID & function name, the initial lookup cost is `O(num_inputs + num_outputs)`.
+    ///    - If an input or output index does not match, the additional lookup cost is `0`.
+    ///    - If an input or output index matches, the additional lookup cost is `O(n)` for `n` restricted arguments with the same index.
+    arguments: IndexMap<Locator<N>, IndexMap<ArgumentLocator, IndexMap<Literal<N>, BlockRange>>>,
```
Comment (evanmarshall) [inline synthesizer/src/restrictions/mod.rs:58]: How would we handle the case where we want to ban an address from doing anything? It seems structurally more difficult than these checks. I imagine our best option right now would be to restriction `credits.aleo/fee` for some address but this is somewhat easily bypassed as the fee payer can be a different signer.

**Review by evanmarshall — DISMISSED** [review]
Overall looks good to me. A couple small changes & a couple issues to file from these changes.

**Review by Meshiest — DISMISSED** [review]

**Comment by Meshiest** [comment]
> But the above scenario is not possible because the node with updated list will not connect and sync blocks from nodes with other lists: [AleoNet/snarkOS#3306](https://github.com/AleoNet/snarkOS/pull/3306)

1. validators sync to block 30k
2. restriction added on block range 29k... for transfers for a certain address
3. all nodes restart with the new restrictions
4. incoming clients (or 1k behind validators) sync up to 29k and start hitting the newly restricted transfers
5. validators that were at 30k have old mappings that do not reflect the restrictions, new clients/lagging validators have different mappings that reflect the restricted transactions

**Comment by vicsn** [comment]
Thank you for explaining @Meshiest . Indeed that sounds like a serious UX issue. 

To further my own understanding and documentation: given that validators frequently lag multiple blocks behind one another even in a healthy network, with the current design "social consensus" would have to enforce restrictions **at least** 10 blocks into the future when any network restart happens which updates the restrictions list. Enforcing such a delay in code is not possible because a validator can't know independently if its behind tip or not, especially because other validators might have already been switched off for the restart.

File: console/network/src/canary_v0.rs (hunk 1 above)
Comment (vicsn) [reply console/network/src/canary_v0.rs:156]: Appreciate the suggestion, but let's leave it as is for now.

File: ledger/benches/bonded_mapping.rs (hunk 5 above)
Comment (vicsn) [reply ledger/benches/bonded_mapping.rs:26]: I think the more style-enforcing pre-commit hooks the better!

File: synthesizer/src/restrictions/mod.rs (hunk 6 above)
Comment (vicsn) [reply synthesizer/src/restrictions/mod.rs:69]: I agree its superfluous, and the UX benefit for certain scenario's might be achievable in many other ways (a restrictions_id script, a checksum utility, ...). But I assume Howard put thought into it and discussed with ANF, so let's leave it unless there is a specific vulnerability.

File: synthesizer/src/restrictions/mod.rs (hunk 7 above)
Comment (vicsn) [reply synthesizer/src/restrictions/mod.rs:58]: Indeed that would be a nice extension. A first step within the current design could be to extend the check against plaintext literal inputs to also plaintext container inputs. But there's a large design space ahead.

**Review by voximity — APPROVED** [review]

**Review by damons — APPROVED** [review]
Approved under duress.  I don't like allowing features which have not been tested.  To me, If it doesn't have a test, it doesn't work.

**Comment by apruden2008** [comment]
> Approved under duress. I don't like allowing features which have not been tested. To me, If it doesn't have a test, it doesn't work.

Working on testing in canarynet now, per discussion with @damons

**Pull request merged**